tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1173,7 +1195,12 @@ unsigned CGMSHLSLRuntime::AddTypeAnnotation(QualType Ty,
   else if (IsHLSLOutputPatchType(Ty))
     return AddTypeAnnotation(GetHLSLOutputPatchElementType(Ty), dxilTypeSys,
                              arrayEltSize);
-  else if (const RecordType *RT = paramTy->getAsStructureType()) {
+  else if (!IsHLSLStructuredBufferType(Ty) && IsHLSLResourceType(Ty)) {
```

Comment:
The `IsHLSLResourceType(Ty)` case existed but was never used because it was after RecordType checks, so all resource object structures just got the same treatment as plain user structures.  Structured buffer must be excluded because we rely on the top-level type annotation and the embedded element type for this resource type, rather than just the type annotation of the element type.